### ISSUE TITLE
Entity pushing is weird/not lag compensated

### ISSUE THREAD

**Issue by MWHunter** [MWHunter/Grim#33]
Sometimes, for optimization, collision doesn't run, which is good!  We should just be able to use the player's real movement with collision for optimization.  I'm currently looking at how collisions actually work in-game.

It messes up badly since the position of mobs aren't lag compensated (do they really need to be?  Can we just lookup nearby collision and add leniency in case there is some?)

Not too sure right now.  It's not a major issue but it definitely will cause issues as everything else becomes more precise.

**Comment by terminalsin** [comment]
Mob collisions will have an offset whether you like it or not. Such is needed to prevent abusing your tick. Instead, you can focus on replicating player position instead, which is much simpler and much more common.

**Comment by DefinePvP** [comment]
I bet you could determine ping if you did enough math on a mob collision.

**Comment by terminalsin** [comment]
> 
> 
> I bet you could determine ping if you did enough math on a mob collision.

Yes, but it would be inaccurate due to the nature of RelativeMove's being sent only a couple of time every second as opposed to once per tick.

**Comment by MWHunter** [comment]
Entity pushing implemented by calculation collisions from lag compensated entity positions and then adding with default 0.04 lenience per collision plus the calculated collision, with an added generic 0.05 per collision subtracted from the offset before collisions are ran.  The number of collisions are an evicting list of size 3 and the average is taken.  The 0.05 offset is also an evicting list with size 3 and taking an average, with the player's hitbox being expanded by 0.5 to find possible collisions.  Might be able to decrease the buffer eventually but I don't care right now.

tl;dr: implemented
